src/RadixDlt.NetworkGateway.PostgresIntegration/Services/VirtualEntityDataProvider.cs

```diff
@@ -0,0 +1,253 @@
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Newtonsoft.Json.Linq;
+using RadixDlt.NetworkGateway.Abstractions;
+using RadixDlt.NetworkGateway.Abstractions.Addressing;
+using RadixDlt.NetworkGateway.Abstractions.Extensions;
+using RadixDlt.NetworkGateway.GatewayApi.Services;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using CoreModel = RadixDlt.CoreApiSdk.Model;
+using GatewayModel = RadixDlt.NetworkGateway.GatewayApiSdk.Model;
+using ToolkitModel = RadixEngineToolkit;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.Services;
+
+public interface IVirtualEntityDataProvider
+{
+    public bool IsVirtualAccountAddress(EntityAddress address);
+
+    public bool IsVirtualIdentityAddress(EntityAddress address);
+
+    (GatewayModel.StateEntityDetailsResponseComponentDetails Details, GatewayModel.EntityMetadataCollection Metadata) GetVirtualEntityData(EntityAddress address);
+}
+
+public class VirtualEntityDataProvider : IVirtualEntityDataProvider
+{
+    private readonly string _accountPackage;
+    private readonly string _identityPackage;
+    private readonly byte _networkId;
+    private readonly byte _secp256k1VirtualAccountDiscriminator;
+    private readonly byte _ed25519VirtualAccountDiscriminator;
+    private readonly byte _secp256k1VirtualIdentityDiscriminator;
+    private readonly byte _ed25519VirtualIdentityDiscriminator;
+    private readonly string _secp256k1SignatureVirtualBadge;
+    private readonly string _ed25519SignatureVirtualBadge;
+    private readonly List<GatewayModel.ComponentEntityRoleAssignmentEntry> _virtualEntityRoleAssignmentEntries;
+
+    public VirtualEntityDataProvider(INetworkConfigurationProvider networkConfigurationProvider)
```

Comment:
hmm i'm not sure but probably we should add entry to changelog about that fix as well.